content/blog/purrr-1-0-0/index.Rmd

```diff
@@ -0,0 +1,485 @@
+-   The `map()` family uses the tidyverse rules for coercion and recycling.
+-   All functions that modify lists handle `NULL` consistently.
+-   We've deprecated functions that aren't related to the core purpose of purr.
+
+### pluck and zero-length vectors
+
+Previously, `pluck()` replaced 0-length vectors with the value of `default`.
+Now `default` is only used for `NULL`s and absent elements:
+
+```{r}
+x <- list(y = list(a = character(), b = NULL))
+x |> pluck("y", "a", .default = NA)
+x |> pluck("y", "b", .default = NA)
+x |> pluck("y", "c", .default = NA)
+```
+
+This also influences the map family because using an integer vector, character vector, or list instead of a function automatically calls `pluck()`:
+
+```{r}
+x <- list(list(1), list(), list(NULL), list(character()))
+x |> map(1, .default = 0) |> str()
+```
+
+We made this change because it makes purrr more consistent with the rest of the tidyverse and it looks like it was a bug in the original implementation of the function.
+
+### Tidyverse consistency
+
+We've tweaked the map family of functions to be more consistent with general tidyverse coercion and recycling rules, as implemented by the [vctrs](https://vctrs.r-lib.org) package.
+`map_lgl()`, `map_int()`, `map_int()`, and `map_dbl()` now follow the same [coercion rules](https://vctrs.r-lib.org/articles/type-size.html#coercing-to-common-type) as vctrs.
+In particular:
+
+-   `map_chr(TRUE, identity)`, `map_chr(0L, identity)`, and `map_chr(1L, identity)` have been deprecated because we believe that converting a logical/integer/double to a character vector is potentially dangerous and should require an explicit coercion.
+
+    ```{r}
+    # previously you could write
+    map_chr(1:4, \(x) x + 1)
+
+    # now you need something like this:
+    map_chr(1:4, \(x) as.character(x + 1))
+    ```
+
+-   `map_int()` requires that the numeric results be close to integers, rather than silently truncating to integers.
+    Compare these two examples:
+
+    ```{r}
+    #| error: true
+    map_int(1:3, \(x) x / 2)
+
+    map_int(1:3, \(x) x * 2)
+    ```
+
+`map2()`, `modify2()`, and `pmap()` use tidyverse recycling rules, which mean that vectors of length 1 are recycled to any size but all other vectors must have the same length.
+This has two major changes:
+
+-   Previously, the presence of a zero-length input generated a zero-length output.
+    Now its recycled using the same rules:
+
+    ```{r}
+    #| error: true
+    map2(1:2, character(), paste)
+
+    # Works because length-1 vector gets recycled to length-0
+    map2(1, character(), paste)
+    ```
+
+-   And now must explicitly recycle vectors that aren't length 1:
+
+    ```{r}
+    #| error: true
+    map2_int(1:4, c(10, 20), `+`)
+
+    map2_int(1:4, rep(c(10, 20), 2), `+`)
+    ```
+
+### Assigning `NULL`
+
+purrr has a number of functions that modify a list: `pluck<-`, `assign_in()`, `modify()`, `modify2()`, `modify_if()`, `modify_at()`, and `list_modify()`.
```

Comment:
```suggestion
purrr has a number of functions that modify a list: `pluck<-()`, `assign_in()`, `modify()`, `modify2()`, `modify_if()`, `modify_at()`, and `list_modify()`.
```

`pluck<-` doesn't generate a link on the blog post. I wonder if it needs `()`? Idk if it is even possible to generate links to assignment functions using hugodown/downlit